storage/store/approvals.go

```diff
@@ -0,0 +1,133 @@
+package store
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/operation"
+)
+
+// ResultApprovals implements persistent storage for result approvals.
+type ResultApprovals struct {
+	db       storage.DB
+	cache    *Cache[flow.Identifier, *flow.ResultApproval]
+	indexing *sync.Mutex // preventing concurrent indexing of approvals
+}
+
+func NewResultApprovals(collector module.CacheMetrics, db storage.DB) *ResultApprovals {
+	store := func(rw storage.ReaderBatchWriter, key flow.Identifier, val *flow.ResultApproval) error {
+		return operation.InsertResultApproval(rw.Writer(), val)
+	}
+
+	retrieve := func(r storage.Reader, approvalID flow.Identifier) (*flow.ResultApproval, error) {
+		var approval flow.ResultApproval
+		err := operation.RetrieveResultApproval(r, approvalID, &approval)
+		return &approval, err
+	}
+
+	res := &ResultApprovals{
+		db: db,
+		cache: newCache[flow.Identifier, *flow.ResultApproval](collector, metrics.ResourceResultApprovals,
+			withLimit[flow.Identifier, *flow.ResultApproval](flow.DefaultTransactionExpiry+100),
+			withStore[flow.Identifier, *flow.ResultApproval](store),
+			withRetrieve[flow.Identifier, *flow.ResultApproval](retrieve)),
+		indexing: new(sync.Mutex),
+	}
+
+	return res
+}
+
+func (r *ResultApprovals) store(rw storage.ReaderBatchWriter, approval *flow.ResultApproval) error {
+	return r.cache.PutTx(rw, approval.ID(), approval)
+}
+
+func (r *ResultApprovals) byID(reader storage.Reader, approvalID flow.Identifier) (*flow.ResultApproval, error) {
+	val, err := r.cache.Get(reader, approvalID)
+	if err != nil {
+		return nil, err
+	}
+	return val, nil
+}
+
+func (r *ResultApprovals) byChunk(reader storage.Reader, resultID flow.Identifier, chunkIndex uint64) (*flow.ResultApproval, error) {
+	var approvalID flow.Identifier
+	err := operation.LookupResultApproval(reader, resultID, chunkIndex, &approvalID)
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup result approval ID: %w", err)
+	}
+	return r.byID(reader, approvalID)
+}
+
+// CAUTION: Caller must acquire `indexing` lock.
+func (r *ResultApprovals) index(rw storage.ReaderBatchWriter, resultID flow.Identifier, chunkIndex uint64, approvalID flow.Identifier) error {
+	var storedApprovalID flow.Identifier
+	err := operation.LookupResultApproval(rw.GlobalReader(), resultID, chunkIndex, &storedApprovalID)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return fmt.Errorf("could not lookup result approval ID: %w", err)
+		}
+
+		// no approval found, index the approval
```

Comment:
should this go after the error conditional completes on line 79?

Reply:
No, line 79 is when we have found an approval indexed by the chunk, in that case, we need to compare if the stored on is the same as the new one. And we should not store if they are different, because we should not allow a different approval for the same chunk.